ksql-cli/src/main/java/io/confluent/ksql/cli/console/JLineReader.java

```diff
@@ -41,14 +42,23 @@
   private final LineReader lineReader;
   private final String prompt;
 
-  // Have to enable event expansion or multi-line parsing won't work, so a quick 'n dirty workaround
-  // will have to do to prevent strings like !! from being expanded by the line reader
-  private static class NoOpExpander extends DefaultExpander {
+  /**
+   * Override the default JLine 'expander' behavior so that history-referencing expressions such
+   * as '!42' or '!!' will only be processed and replaced if they occur at the beginning of an
+   * input line, even if the input line spans multiple terminal lines using the '\' separator.
+   */
+  private static class KsqlExpander extends DefaultExpander {
 
     @Override
     public String expandHistory(final History history, final String line) {
-      return line;
+      if (line.startsWith("!") || line.startsWith("^")) {
```

Comment:
is line trimmed? What if there's leading spaces?

Reply:
line is not trimmed, deliberately ignoring leading whitespace. i wanted to force the special handling of these history codes to only work if you deliberately typed them as first thing on a new line (that's not at all the default way JLine wants to do this). My thinking here was that lines with leading whitespace might be intended as some kind of continuation/indentation of a previous query and hence should be excluded.